p2p/transport/webrtc/internal/util.go

```diff
@@ -0,0 +1,87 @@
+package internal
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/libp2p/go-libp2p/p2p/transport/webrtc/internal/encoding"
+	ma "github.com/multiformats/go-multiaddr"
+	"github.com/multiformats/go-multibase"
+	mh "github.com/multiformats/go-multihash"
+	"github.com/pion/datachannel"
+	"github.com/pion/webrtc/v3"
+)
+
+func DecodeRemoteFingerprint(maddr ma.Multiaddr) (*mh.DecodedMultihash, error) {
+	remoteFingerprintMultibase, err := maddr.ValueForProtocol(ma.P_CERTHASH)
+	if err != nil {
+		return nil, err
+	}
+	_, data, err := multibase.Decode(remoteFingerprintMultibase)
+	if err != nil {
+		return nil, err
+	}
+	return mh.Decode(data)
+}
+
+func EncodeDTLSFingerprint(fp webrtc.DTLSFingerprint) (string, error) {
+	digest, err := encoding.DecodeInterpersedHexFromASCIIString(fp.Value)
+	if err != nil {
+		return "", err
+	}
+	encoded, err := mh.Encode(digest, mh.SHA2_256)
+	if err != nil {
+		return "", err
+	}
+	return multibase.Encode(multibase.Base64url, encoded)
+}
+
+// only use this if the datachannels are detached, since the OnOpen callback
+// will be called immediately. Only use after the peerconnection is open.
+// The context should close if the peerconnection underlying the datachannel
+// is closed.
+func GetDetachedChannel(ctx context.Context, dc *webrtc.DataChannel) (rwc datachannel.ReadWriteCloser, err error) {
+	done := make(chan struct{})
+	dc.OnOpen(func() {
+		defer close(done)
+		rwc, err = dc.Detach()
+	})
+	// this is safe since for detached datachannels, the peerconnection runs the onOpen
+	// callback immediately if the SCTP transport is also connected.
+	select {
+	case <-done:
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+	return
+}
+
+func AwaitPeerConnectionOpen(ufrag string, pc *webrtc.PeerConnection) <-chan error {
+	errC := make(chan error)
+	var once sync.Once
+	pc.OnConnectionStateChange(func(state webrtc.PeerConnectionState) {
+		switch state {
+		case webrtc.PeerConnectionStateConnected:
+			once.Do(func() { close(errC) })
+		case webrtc.PeerConnectionStateFailed:
+			once.Do(func() {
```

Comment:
In this function, we are only concerned with the PeerConnection moving to the `Connected` state, at which point we close the error channel and the function returns. Once this returns, a `Connection` is set up which changes the `OnPeerConnectionStateChange` callback to manage the connection transitioning to different states. The `once` is to ensure we do not send on a closed channel.

If a connection moves to a failed state (which will always occur after a delay because of the `DisconnectedTimeout` and `FailedTimeout` set in the PeerConnection configuration) after this function returns but before the noise handshake, the noise handshake will fail, and the `Connection` object wrapping the peerconnection will also try and close the connection.